shortfin/python/shortfin_apps/sd/components/builders.py

```diff
@@ -0,0 +1,164 @@
+        if model_params.base_model_name.lower() == "sdxl"
+        else [model_params.base_model_name]
+    )
+    mod_names = {
+        "clip": "clip",
+        "unet": "punet" if model_params.use_i8_punet else "unet",
+        "scheduler": model_params.scheduler_id + "Scheduler",
+        "vae": "vae",
+    }
+    for mod, modname in mod_names.items():
+        ord_params = [
+            base,
+            [modname],
+        ]
+        bsizes = []
+        for bs in getattr(model_params, f"{mod}_batch_sizes", [1]):
+            bsizes.extend([f"bs{bs}"])
+        ord_params.extend([bsizes])
+        if mod in ["unet", "clip"]:
+            ord_params.extend([[str(model_params.max_seq_len)]])
+        if mod in ["unet", "vae", "scheduler"]:
+            dims = []
+            for dim_pair in model_params.dims:
+                dim_pair_str = [str(d) for d in dim_pair]
+                dims.extend(["x".join(dim_pair_str)])
+            ord_params.extend([dims])
+        if mod == "scheduler":
+            dtype_str = dtype_to_filetag[model_params.unet_dtype]
+        elif mod != "unet":
+            dtype_str = dtype_to_filetag[
+                getattr(model_params, f"{mod}_dtype", sfnp.float16)
+            ]
+        else:
+            dtype_str = (
+                "i8"
+                if model_params.use_i8_punet
+                else dtype_to_filetag[model_params.unet_dtype]
+            )
+        ord_params.extend([[dtype_str]])
+        for x in list(itertools.product(*ord_params)):
+            file_stems.extend(["_".join(x)])
+    return file_stems
+
+
+def get_url_map(filenames: list[str], bucket: str):
+    file_map = {}
+    for filename in filenames:
+        file_map[filename] = f"{bucket}{filename}"
+    return file_map
+
+
+@entrypoint(description="Retreives a set of SDXL submodels.")
+def sdxl(
+    model_json=cl_arg(
+        "model_json",
+        default=default_config_json,
+        help="Local config filepath",
+    ),
+    target=cl_arg(
+        "target",
+        default="gfx942",
+        help="IREE HIP target arch",
+    ),
+    splat=cl_arg("splat", default=False, help="Download empty weights (for testing)"),
+):
+    model_params = ModelParams.load_json(model_json)
+    mlir_bucket = SDXL_BUCKET + "mlir/"
+    mlir_filenames = get_mlir_filenames(model_params)
+    mlir_urls = get_url_map(mlir_filenames, mlir_bucket)
+    for f, url in mlir_urls.items():
+        fetch_http(name=f, url=url)
+    vmfb_bucket = SDXL_BUCKET + "vmfbs/"
+    vmfb_filenames = get_vmfb_filenames(model_params, target=target)
+    vmfb_urls = get_url_map(vmfb_filenames, vmfb_bucket)
+    for f, url in vmfb_urls.items():
+        fetch_http(name=f, url=url)
+    params_filenames = get_params_filenames(model_params, splat)
+    params_urls = get_url_map(params_filenames, SDXL_WEIGHTS_BUCKET)
+    ctx = executor.BuildContext.current()
+    for f, url in params_urls.items():
+        out_file = os.path.join(ctx.executor.output_dir, f)
```

Comment:
This isn't very robust. Should have a md5sum checklist fetched from the server if downloads enabled, and compare with local checklist to determine which, if any, artifacts need updating.

Reply:
Yeah, I hadn't yet gotten to stamp and change detection... Will in a bit. 

Do you already have file hashes stored in the bucket somewhere?

Reply:
Not yet. Right now, the vmfbs are always downloaded from a bucket versioned by date only.

Reply:
That can work. You basically need something to derive a stamp value from. That can come from some part of the URL.

Reply:
For mammoth files, a manual version of some kind can be best anyway: it can take a long time to compute a hash of such things

Reply:
Does it seem too heavyweight to keep a md5sums.json in each bucket, and have the builder generate and keep a local set of hashes for its outputs? That way we can filter  exactly what's needed before doing fetch_http?

Reply:
Yeah, that's the basic mechanism. We wouldn't actually compute the hash in the builder in typical use. Instead, you would tell it how to get the stamp artifact (ie. Some fixed string, a hash file, etc). If a hash file, we compute a running hash only during download and store the result, erroring if it mismatches. But just an opaque stamp value drives the up-to-date check. 

It's better for everyone if such artifacts are in write once storage (ie. The same URL produces the same content for all of time). Then the stamp is just the url, and any has checking is just for verifying the integrity of the transfer.